### PR TITLE
Add support for pooled sharded queues

### PR DESCRIPTION
**Pull request by lulf**
Add another field 'partitions' to address plans, which will cause an
address to be allocated to multiple brokers. The broker credits will be
divided by the number of partitions, and the address will be created on
multiple brokers with a fraction of the credits on each.

Removes use of annotations for broker allocations as it is now stored in
status. Rewrite logic that depended on annotations to always use broker
status.

@rgodfrey @k-wall @vbusch @robshelly 

We've talked about this for some time, being able to shard 'small' queues across multiple brokers. With this PR that is now possible, and it was actually small modification to the address pland and scheduleAddress logic to get it to work. The address plan gets a field `partitions` that is default 1 if not set. If set to > 1, the usage and scheduling will use fractions of the credits spread across the different brokers. An alternative would be to multiply the credits by the number of partitions, but that can easily be changed in the implementation. 

There also might be some pitfalls as this can potentially be used on topics as well, with a potential consequence that subscribers on one broker shard does not get messages sent to a different shard. For this reason we should probably ignore the partitions setting for topics and subscriptions for now.

I'm also open for renaming `partitions` to something else, I just felt it would map well to a Kafka topic partition, but its a slightly different thing perhaps.

### CODE REVIEW

**Comment by lulf** [comment]
@enmasse-ci run tests

**Comment by rgodfrey** [comment]
@lulf without looking at the code... how hard would it be to modify this further so that it works for all queues, not just "small" queues.  Rather than partitions defaulting to 1 have partitions default to the smallest integer at or above the credit

**Comment by lulf** [comment]
That is a very minor change. It would however override the “sharded queue” setup that exists today, and I  would need to test that migration would work as expected. For sharded topics we would not want to do this I guess for the same reasons as for pooled topics

**Comment by lulf** [comment]
@enmasse-ci run tests testcase=standard.QueueTest

**Comment by lulf** [comment]
@enmasse-ci run tests testcase=PlansTest

**Comment by rgodfrey** [comment]
@lulf agreed - not something for topics, but having a common model for all queues ignoring size would be nice.

**Comment by lulf** [comment]
@enmasse-ci run tests profile=systemtests-upgrade

**Comment by lulf** [comment]
@enmasse-ci run tests

**Comment by lulf** [comment]
@enmasse-ci run tests testcase=io.enmasse.systemtest.standard.mqtt.SessionTest#previousPersistentSessionDiscarded

**Comment by k-wall** [comment]
Looks okay, but I need to spend time playing with the feature before clicking Approve.

**Comment by k-wall** [comment]
I see a circumstance when then accounting reported in the standard-controller-logs looks off.

I have changed `standard-small-queue` to have partitions 2.  I start off with the plan declaring resources:

```
  resources:
    broker: 0.01
    router: 0.001
```

I have two queues using the plan.

standard-controller reports looks good:

```
2019-03-30T20:50:41.617Z INFO  [AddressController] Usage: {router={all={used=0.002, needed=1}}, broker={broker-er9ae506z8-1k4z={used=0.01, needed=1}, broker-er9ae506z8-kppt={used=0.01, needed=1}}}
2019-03-30T20:50:41.617Z INFO  [AddressController] Needed: {router={all={used=0.002, needed=1}}, broker={broker-er9ae506z8-1k4z={used=0.01, needed=1}, broker-er9ae506z8-kppt={used=0.01, needed=1}}}
```

I now double the broker resources.

```
  resources:
    broker: 0.02
    router: 0.001
```

standard-controller reports look off.   The `needed` line reports twice the amount of broker required.  

```
2019-03-30T20:51:12.174Z INFO  [AddressController] Total: 2, Active: 2, Configuring: 0, Pending: 0, Terminating: 0, Failed: 0
2019-03-30T20:51:12.175Z INFO  [AddressController] Usage: {router={all={used=0.002, needed=1}}, broker={broker-er9ae506z8-1k4z={used=0.02, needed=1}, broker-er9ae506z8-kppt={used=0.02, needed=1}}}
2019-03-30T20:51:12.175Z INFO  [AddressController] Needed: {router={all={used=0.004, needed=1}}, broker={broker-er9ae506z8-1k4z={used=0.04, needed=1}, broker-er9ae506z8-kppt={used=0.04, needed=1}}}

```

**Comment by k-wall** [comment]
Also, I notice if once queues exist, I change the partition count in their plan, decreases in the number of partitions work as expected (I see the standard controller await the shards being empty and then the redundant brokers being shutdown), however, scaling up the partitions is seemingly ignored.

**Comment by k-wall** [comment]
I also think docs would need to be enhanced: https://enmasse.io/documentation/master/openshift/#con-address-plans-messaging

**Comment by lulf** [comment]
> I see a circumstance when then accounting reported in the standard-controller-logs looks off.
> 
> I have changed `standard-small-queue` to have partitions 2. I start off with the plan declaring resources:
> 
> ```
>   resources:
>     broker: 0.01
>     router: 0.001
> ```
> 
> I have two queues using the plan.
> 
> standard-controller reports looks good:
> 
> ```
> 2019-03-30T20:50:41.617Z INFO  [AddressController] Usage: {router={all={used=0.002, needed=1}}, broker={broker-er9ae506z8-1k4z={used=0.01, needed=1}, broker-er9ae506z8-kppt={used=0.01, needed=1}}}
> 2019-03-30T20:50:41.617Z INFO  [AddressController] Needed: {router={all={used=0.002, needed=1}}, broker={broker-er9ae506z8-1k4z={used=0.01, needed=1}, broker-er9ae506z8-kppt={used=0.01, needed=1}}}
> ```
> 
> I now double the broker resources.
> 
> ```
>   resources:
>     broker: 0.02
>     router: 0.001
> ```
> 
> standard-controller reports look off. The `needed` line reports twice the amount of broker required.
> 
> ```
> 2019-03-30T20:51:12.174Z INFO  [AddressController] Total: 2, Active: 2, Configuring: 0, Pending: 0, Terminating: 0, Failed: 0
> 2019-03-30T20:51:12.175Z INFO  [AddressController] Usage: {router={all={used=0.002, needed=1}}, broker={broker-er9ae506z8-1k4z={used=0.02, needed=1}, broker-er9ae506z8-kppt={used=0.02, needed=1}}}
> 2019-03-30T20:51:12.175Z INFO  [AddressController] Needed: {router={all={used=0.004, needed=1}}, broker={broker-er9ae506z8-1k4z={used=0.04, needed=1}, broker-er9ae506z8-kppt={used=0.04, needed=1}}}
> ```

I would expect the amount of broker to doubled if you're doubling the number of brokers required. I would expect usage to correct itself once the updated plan was actually applied. This should get reflected in address.status.

**Comment by lulf** [comment]
@enmasse-ci run tests testcase=plans.PlansTest

**Comment by lulf** [comment]
@enmasse-ci run tests testcase=plans.PlansTest

**Comment by lulf** [comment]
@enmasse-ci run tests testcase=plans.PlansTest#testScalePlanPartitions

**Comment by lulf** [comment]
@enmasse-ci run tests testcase=plans.PlansTest#testScalePlanPartitions

**Comment by lulf** [comment]
@enmasse-ci run tests testcase=plans.PlansTest#testScalePlanPartitions

**Comment by lulf** [comment]
@enmasse-ci run tests testcase=plans.PlansTest#testScalePlanPartitions

**Comment by lulf** [comment]
@k-wall the issues should be addressed now

**Comment by k-wall** [comment]
> > standard-controller reports look off. The `needed` line reports twice the amount of broker required.
> > ```
> > 2019-03-30T20:51:12.174Z INFO  [AddressController] Total: 2, Active: 2, Configuring: 0, Pending: 0, Terminating: 0, Failed: 0
> > 2019-03-30T20:51:12.175Z INFO  [AddressController] Usage: {router={all={used=0.002, needed=1}}, broker={broker-er9ae506z8-1k4z={used=0.02, needed=1}, broker-er9ae506z8-kppt={used=0.02, needed=1}}}
> > 2019-03-30T20:51:12.175Z INFO  [AddressController] Needed: {router={all={used=0.004, needed=1}}, broker={broker-er9ae506z8-1k4z={used=0.04, needed=1}, broker-er9ae506z8-kppt={used=0.04, needed=1}}}
> > ```
> 
> @k-wall I would expect the amount of broker to doubled if you're doubling the number of brokers required. I would expect usage to correct itself once the updated plan was actually applied. This should get reflected in address.status.

@lulf I don't understand.  If my my plan says the queue needs 0.02 of a broker and I have two broker, then I expect to see 0.01 used and 0.01% needed on each.  Am I thinking about this in the wrong way?

**Comment by lulf** [comment]
> > > standard-controller reports look off. The `needed` line reports twice the amount of broker required.
> > > ```
> > > 2019-03-30T20:51:12.174Z INFO  [AddressController] Total: 2, Active: 2, Configuring: 0, Pending: 0, Terminating: 0, Failed: 0
> > > 2019-03-30T20:51:12.175Z INFO  [AddressController] Usage: {router={all={used=0.002, needed=1}}, broker={broker-er9ae506z8-1k4z={used=0.02, needed=1}, broker-er9ae506z8-kppt={used=0.02, needed=1}}}
> > > 2019-03-30T20:51:12.175Z INFO  [AddressController] Needed: {router={all={used=0.004, needed=1}}, broker={broker-er9ae506z8-1k4z={used=0.04, needed=1}, broker-er9ae506z8-kppt={used=0.04, needed=1}}}
> > > ```
> > 
> > 
> > @k-wall I would expect the amount of broker to doubled if you're doubling the number of brokers required. I would expect usage to correct itself once the updated plan was actually applied. This should get reflected in address.status.
> 
> @lulf I don't understand. If my my plan says the queue needs 0.02 of a broker and I have two broker, then I expect to see 0.01 used and 0.01% needed on each. Am I thinking about this in the wrong way?

@k-wall Sorry, this was wrong reasoning on my part. If you have credits 0.2 with 1 partition, you get 0.2 on 1 broker. If you have 0.2 with 2 partitions, you get 0.1 on 2 brokers. I have addresses this issue in the latest fixes.

File: documentation/modules/con-address-plans.adoc [hunk 1]
```diff
@@ -26,6 +26,11 @@ The following fields are required:
 
 A single router can support five instances of addresses and broker can support three instances of addresses with this plan. If the number of addresses with this plan increases to four, another broker is created. If it increases further to six, another router is created as well.
 
+In the `standard` address space, address plans for the address type `queue` also supports a field
+`partitions` which allows a queue to be sharded accross multiple brokers for HA and performance
+benefits, at the cost of loosing message ordering. Specifying an amount of `broker` resource above
```
Comment (k-wall) [inline documentation/modules/con-address-plans.adoc]: type: losing

**Comment by k-wall** [comment]
If a queue's partitions exceeds the address space plan's broker resource limits, I would expect the address to go into a quota exceed case, but I am not seeing this.  I see the number of broker pods go above `resourceLimits.broker`.

```
NAME                ADDRESS    ADDRESSSPACE   TYPE      PLAN                   READY     PHASE     AGE       STATUS
standard.myqueue1   myqueue1   standard       queue     standard-small-queue   true      Active    26m
standard.myqueue2   myqueue2   standard       queue     standard-large-queue   false     Pending   6m        Quota exceeded
```
standard-small-queue has partitions 5.  standard-large-queue is not partitioned.

```
[keith@rhfed-localdomain enmasse]$ oc get pods
NAME                                        READY     STATUS    RESTARTS   AGE
address-space-controller-577df96688-6vlwp   1/1       Running   0          30m
admin.ryaepdxnud-9c5b55664-qs9fz            2/2       Running   0          30m
api-server-7698957c6f-2sjrm                 1/1       Running   0          30m
broker-ryaepdxnud-64ia-0                    1/1       Running   0          12m
broker-ryaepdxnud-9dgq-0                    1/1       Running   0          13m
broker-ryaepdxnud-ulyh-0                    1/1       Running   0          11m
broker-ryaepdxnud-uxjh-0                    1/1       Running   0          13m
broker-ryaepdxnud-wksp-0                    1/1       Running   0          13m
enmasse-operator-6cddbd9bff-qwrgc           1/1       Running   0          30m
qdrouterd-ryaepdxnud-0                      1/1       Running   0          30m
qdrouterd-ryaepdxnud-1                      1/1       Running   0          28m
standard-authservice-dbb46d64c-k2grx        1/1       Running   0          30m

```

**Comment by lulf** [comment]
> If a queue's partitions exceeds the address space plan's broker resource limits, I would expect the address to go into a quota exceeded state, but I am not seeing this. I see the number of broker pods go above `resourceLimits.broker`.
> 
> ```
> NAME                ADDRESS    ADDRESSSPACE   TYPE      PLAN                   READY     PHASE     AGE       STATUS
> standard.myqueue1   myqueue1   standard       queue     standard-small-queue   true      Active    26m
> standard.myqueue2   myqueue2   standard       queue     standard-large-queue   false     Pending   6m        Quota exceeded
> ```


Are you modifying the plan after you've deployed the queues or before? I'm asking because I'm trying to reproduce this in a unit test without success yet.

**Comment by lulf** [comment]
@enmasse-ci run tests testcase=plans.PlansTest#testScalePlanPartitions

**Comment by lulf** [comment]
@enmasse-ci run tests testcase=plans.PlansTest#testScalePlanPartitions

**Comment by lulf** [comment]
@enmasse-ci run tests testcase=plans.PlansTest#testScalePlanPartitions

**Comment by lulf** [comment]
@enmasse-ci run tests testcase=plans.PlansTest#testScalePlanPartitions

**Pull request merged**